h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -447,18 +449,32 @@ public void modelFailed(Model failedModel) {
 
     /** All visited hyper params permutations, including the current one. */
     private List<int[]> _visitedPermutations = new ArrayList<>();
+    private int _numberOfNonSkippedPermutations = 0;
     private Set<Integer> _visitedPermutationHashes = new LinkedHashSet<>(); // for fast dupe lookup
 
+    ArrayList<PermutationFilterFunction> _filterFunctions;
+
     public RandomDiscreteValueWalker(MP params,
-                                     Map<String, Object[]> hyperParams,
+                                     Map<String, Object[]> hyperParamsGrid,
                                      ModelParametersBuilderFactory<MP> paramsBuilderFactory,
-                                     HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria search_criteria) {
-      super(params, hyperParams, paramsBuilderFactory, search_criteria);
+                                     HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria searchCriteria,
+                                     ArrayList<PermutationFilterFunction> filterFunctions) {
```

Comment:
@honzasterba I don't agree that semantic of collection of the filters is not clear. This concept is everywhere: when you filter your tap water - each filter is filtering out something specific( depending on its purpose) and results of filtering might overlap as well. But I agree that we can add description that it is AnyMatch collection of functions.
In this particular case we need (at least) two different types of filter functions - one is stateful and another one is not. 
If we care about clarity then combining different types of filters into one function would be even more difficult to reason about. It will look simpler from first glance of API user's perspective - as we kind of don't care what will be inside of the function, but implementation of this function will be less convenient. In my approach I already provide explicitly tools to specify two different types of filtering functions.

Reply:
I dot know how to respond to this. @michalkurka, please help

Reply:
@honzasterba I think one of your points is to refactor the state out of RandomDiscreteValueWalker and move the responsibility somewhere else. I think that is a good idea. Lets see what @deil87 thinks and then we can move to this topic. I think solving one would help here as well.

As the bare minimum we should probably document what the list of functions does / what is the behavior.

Reply:
@michalkurka 
1) As I previously mentioned here https://github.com/h2oai/h2o-3/pull/4040#discussion_r345995929 moving logic outside will result in unnecessary `Model.Parameters` instantiations because current API returns `MP` generic type.
2) Regarding separation of concerns https://github.com/h2oai/h2o-3/pull/4040#discussion_r347614164 . Existing version of `RandomDiscreteValueWalker` actually doing filtering inside its iterator,  just this filter function is simpler and does not require parametrisation from the user. If we wanted to move this concern outside of the walkers then we would have to make `RandomDiscreteValueWalker` to return any permutation it steps into as a `Model.Parameters` instance  and only then discard it in the outside Wrapper that will be keeping track of visited permutations. But existing implementation is not doing it like this - it filters duplicates inside iterator.  Speaking of concerns - class with name `RandomDiscreteValueWalker` should not care about duplicates as randomness does not care about repetitions. As soon as we all agree on this point it would be easier to move on. My version is just following existing approach. 

I'm all for separation of concern but I'm also for consistency: If we want to move this concern outside then I would move deduplication logic from `RandomDiscreteValueWalker` as well.

> As the bare minimum we should probably document what the list of functions does / what is the behavior.

sure, I will be glad to add more explanations to existing ones